python/ray/ml/examples/pytorch_geometric/distributed_reddit_example.py

```diff
@@ -0,0 +1,178 @@
+# Adapted from https://github.com/pyg-team/pytorch_geometric/blob/master/examples
+# /multi_gpu/distributed_sampling.py.
+
+import os
+import argparse
+from filelock import FileLock
+
+import torch
+import torch.nn.functional as F
+
+from torch_geometric.datasets import Reddit
+from torch_geometric.loader import NeighborSampler
+from torch_geometric.nn import SAGEConv
+
+from ray import train
+from ray.ml.train.integrations.torch import TorchTrainer
+
+
+class SAGE(torch.nn.Module):
+    def __init__(self, in_channels, hidden_channels, out_channels, num_layers=2):
+        super().__init__()
+        self.num_layers = num_layers
+
+        self.convs = torch.nn.ModuleList()
+        self.convs.append(SAGEConv(in_channels, hidden_channels))
+        for _ in range(self.num_layers - 2):
+            self.convs.append(SAGEConv(hidden_channels, hidden_channels))
+        self.convs.append(SAGEConv(hidden_channels, out_channels))
+
+    def forward(self, x, adjs):
+        for i, (edge_index, _, size) in enumerate(adjs):
+            x_target = x[: size[1]]  # Target nodes are always placed first.
+            x = self.convs[i]((x, x_target), edge_index)
+            if i != self.num_layers - 1:
+                x = F.relu(x)
+                x = F.dropout(x, p=0.5, training=self.training)
+        return x.log_softmax(dim=-1)
+
+    @torch.no_grad()
+    def inference(self, x_all, subgraph_loader):
+        for i in range(self.num_layers):
+            xs = []
+            for batch_size, n_id, adj in subgraph_loader:
+                edge_index, _, size = adj
+                x = x_all[n_id].to(train.torch.get_device())
+                x_target = x[: size[1]]
+                x = self.convs[i]((x, x_target), edge_index)
+                if i != self.num_layers - 1:
+                    x = F.relu(x)
+                xs.append(x.cpu())
+
+            x_all = torch.cat(xs, dim=0)
```

Comment:
looks a bit weird to me. I think I am just clueless.
if we overwrite the entire x_all here, we will only have features for the nodes that we scored with the last layer.
it feels more appropriate to me to update the weights of xs in x_all, not simply "x_all = ..." ??

Reply:
This is just the inference code so no weights updating. I think this works the same way as a standard feed-forward neural network. We only want the output of the last layer, and we don't care about the hidden states during inference.

Reply:
ok, I understand this now. subgraph_loader actually samples a subgraph for every single node in the graph.
so if there are n nodes in the graph, the inner loop will run n times. each time, we essentially aggregating data from all neighboring nodes to this specific node.
so at the end, torch.cat(xs) will give us a new updated graph, since xs will contain data for every single node at the point.
interesting design.